src/main/java/au/org/ala/biocache/service/DownloadService.java

```diff
@@ -129,80 +152,175 @@
+                                // These specialised control threads get a higher thread priority than 
+                                // the default control thread that would otherwise accept all downloads
+                                // This is not vital, as there are polling delay differences to also account
+                                // for the difference, but in the event of having to wake threads up by priority,
+                                // this tells the JVM that these are more useful than the default thread.
+                                nextThread.setPriority(Thread.NORM_PRIORITY);
+                                runningDownloadControllers.add(nextThread);
+                                runningDownloadControlRunnables.add(nextRunnable);
+                                nextThread.start();
                             }
-                            new Thread(new DownloadThread(maxRecords, dt)).start();
-                            threads--;
+                        } catch (Exception e) {
+                            logger.error("failed to create all extra offline download threads for concurrent.downloads.extra=" + concurrentDownloadsExtra, e);
                         }
+                    } finally {
+                        initialisationLatch.countDown();
+                    }
+                }
+            }.start();
+        }
+    }
+    
+    /**
+     * Ensures that all of the download threads are given a chance to shutdown cleanly using thread interrupts.
+     */
+    @Override
+    public void onApplicationEvent(ContextClosedEvent event) {
+        if(closed.compareAndSet(false, true)) {
+            afterInitialisation();
+            try {
+                // Stop more downloads from being added by shutting down additions to the persistent queue
+                persistentQueueDAO.shutdown();
+            }
+            finally {
+                DownloadControlThread nextToCloseRunnable = null;
+                // Interrupt all of the download control threads
+                while((nextToCloseRunnable  = runningDownloadControlRunnables.poll()) != null) {
+                    nextToCloseRunnable.shutdown();
+                }
+                
+                Thread nextToCloseThread = null;
+                List<Thread> toJoinThreads = new ArrayList<>();
+                while((nextToCloseThread = runningDownloadControllers.poll()) != null) {
+                    if(nextToCloseThread.isAlive()) {
+                        nextToCloseThread.interrupt();
+                        toJoinThreads.add(nextToCloseThread);
+                    }
+                }
+                
+                // Give the download control threads a few seconds to cleanup before returning from this method
+                for(Thread nextToJoinThread : toJoinThreads) {
+                    try {
+                        nextToJoinThread.join(5000);
```

Comment:
`5000 * toJoinThreads.size()` is conceivably large . Replace with a single 5s sleep.